### PR TITLE
 Enable download nightly JDKs from AdoptOpenJDK

### LINKED ISSUE #1: Download JDKs from AdoptOpenJDK nightly builds ( EA)
The AdoptOpenJDK API provides a way to consume AdoptOpenJDK releases and nightly builds. Current only download releases(GA) JDKs is supported. Nice to also support download nightly (EA) JDKs.

https://github.com/AdoptOpenJDK/install-jdk/blob/master/src/installer.ts#L66

Which would be consistent with AdoptOpenJDK/openjdk-tests story. I.E. sdk source: nightly, releases, customized, upstream.

nightly --> AdoptOpenJDK V3 API  EA
releases --> AdoptOpenJDK V3 API  GA
customized --> A custom source location of a JDK. A local directory, a compressed file, or a url.
upstream --> AdoptOpenJDK/build-jdk action or could be included in customized as a local 
                        compressed file

### PR DESCRIPTION
**Pull request by sophia-guo**
Implement  #1 

Signed-off-by: Sophia Guo <sophiag@ca.ibm.com>

### CODE REVIEW

**Comment by sophia-guo** [comment]
For the case of the sdk source: upstream
AdoptOpenJDK/build-jdk action generate a JDK tar file, which is a local file of caller. It could be treated as source: a local file when using install-jdk.

**Review by TheMrMilchmann — COMMENTED** [review]
Hey, thanks for your work. The implementation looks fine but I don't like overloading the `source` property. Instead, I would like to expose a new property called `buildType` which is set to "release" by default and only used if the default source is used (i.e. `source` is not set). This would 1) avoid potential clashes between build types and directories ("nightly" or "release" could, in theory, be local directories with a valid JDK) and 2) be more future proof should there ever be a need for additional build types. (What comes to my mind here, is potential support for OpenJDK project-specific EA builds at some point.)

If you agree, I will take care of editing and merging this.

File: action.yml [hunk 1]
```diff
@@ -9,8 +9,9 @@ inputs:
     required: false
     default: 'x64'
   source:
-    description: 'A custom JDK location. May either be a local archive, folder, or URL.'
+    description: 'A custom JDK location. May either be [nighlty, releases(default)] from AdoptOpenJDK or be a local archive, folder.'
```
Comment (smlambert) [inline action.yml]: typo: nighlty -> nightly

**Review by smlambert — COMMENTED** [review]
A bit of background for @TheMrMilchmann, there are already several SDK_RESOURCE types that AQA supports, as seen in our Jenkins pipelines, there is an environment variable / pipeline parameter called SDK_RESOURCE, that currently can be set to one of 4 things:
- upstream (which in a Jenkins environment means go pull the binary from another Jenkins job)
- nightly (which pulls a nightly build from AdoptOpenJDK)
- releases (which pulls a release build from AdoptOpenJDK)
- customized (which means the developer can provide a URL that is not one of the above where a binary can be fetched from, which could be a URL to an EA build resides)

Screen Shot 2020-03-23 at 5 06 49 PM

It will be good to continue to have the possible values of SDK_RESOURCE mean the same across environments (Jenkins, Github actions, etc).

The install-jdk action provides a way to fetch nightly and release binaries, where a build-jdk action would provide the equivalent of an upstream binary.

**Comment by TheMrMilchmann** [comment]
Thank you for the additional context!

While I don't agree that this level of continuity is required or even desirable here since this action has potentially much broader applicability than just AQA, it seems reasonable in this case.

I suppose we could avoid potential conflicts between build types and local files by requiring a prefixed [file URI](https://en.wikipedia.org/wiki/File_URI_scheme) for local files. Any objections?

**Comment by sophia-guo** [comment]
> I suppose we could avoid potential conflicts between build types and local files by requiring a prefixed file URI for local files. Any objections?

Does the prefixed file URI mean the input has to be an absolute path？That would be a little bit more work for user.

**Comment by TheMrMilchmann** [comment]
> Does the prefixed file URI mean the input has to be an absolute path?

You're right. I thought file URIs supported relative paths, but they do not.
I currently see two possibilities:

1. Defy the standard and support relative file URIs. (This sounds like a terrible idea to me, but it would be possible.)
2. Add another parameter called `sourceType` with the supported values:
    - `buildType` - `source` is treated as build type
    - `file` - `source` is treated as file path
    - `url` - `source` is treated as HTTP or HTTPs URL.
    - `detect` (default) - Automatically use `url` for "http"- and "https"-prefixed inputs, else check if `source` is a valid file path. Consider `source` to be a `file` or `buildType` depending on whether such a file exists.

To me, option 2. seems to be the most logical solution as it would support all common cases out of the box, while also providing configuration options in situations where the default causes issues. Further, in most cases, the additional parameter can safely be ignored.

**Comment by sophia-guo** [comment]
> Consider source to be a file or buildType depending on whether such a file exists.

The main point is to `avoid potential clashes between build types and directories ("nightly" or "release" could, in theory, be local directories with a valid JDK)`?

In that case adding another parameter call local_jdkfile (default is empty) should be enough, which  is similar to the implementation of `setup-java`. If it is non-empty `source` as `nightly` or `release` or any other value will be ignored. That is user input( customized) is most important. 

>  should there ever be a need for additional build types. (What comes to my mind here, is potential support for OpenJDK project

source could be updated with `ea` if `ea` build is available.

IMO option 2 in https://github.com/AdoptOpenJDK/install-jdk/pull/4#issuecomment-608385500 is clear but complicated especially from user experience.  The values of `sourceType` and `source` are interdependent.

**Comment by TheMrMilchmann** [comment]
> In that case adding another parameter call local_jdkfile (default is empty) should be enough, which is similar to the implementation of setup-java. If it is non-empty source as nightly or release or any other value will be ignored. That is user input( customized) is most important.

As this project is based on setup-java, I'm aware of the `jdkFile` parameter used there. I deliberately decided to replace it with the `source` parameter to support not only local files but also URLs in a single place. Thus, as it stands now, the parameter is already overloaded with multiple different meanings. Currently, how the parameter should be treated is inferred from its value already.

When splitting off local file handling by adding another parameter, the `source` parameter would still remain overloaded and we end up in a situation where we would declare "buildType" sources and URL sources in `source` but file sources in `jdkFile`. I think that adding another overload to `source` (and an optional parameter to resolve conflicts manually) is logically consistent and more in line with the design of install-jdk (and similar to how "target" environment variables are currently treated).

Additionally, it is worth noting that removing file support from the `source` parameter would be a breaking change which (unlike #4) does not fix a major usability problem.

> IMO option 2 in #4 (comment) is clear but complicated especially from user experience. The values of sourceType and source are interdependent.

For most users, `sourceType` would never matter since it would be automatically inferred correctly in most cases. (For now, it would only have to be set explicitly to `buildType` in case a "releases" or "nightly" file exists at the repository root.)

**Comment by sophia-guo** [comment]
@TheMrMilchmann     Could you please do the update and merge it according to option 2 in https://github.com/AdoptOpenJDK/install-jdk/pull/4#issuecomment-608385500  or create a new PR to handle this? I think the most important | urgent step is to enable `nightly` query for now.  So we could run run-aqa action against nightly JDKs easily.  Thanks!

**Comment by TheMrMilchmann** [comment]
Thanks for your work on this and your input! I quickly created #7 based on this PR to solve this issue as specified in option 2 and I'll come back to this later today (probably in ~10 hours) to do some more testing and to merge the changes.

**Pull request closed without merging**